File: examples/hello-world/gabriel-greeting-dart/api/cli_test.dart
import 'dart:convert';

import 'package:test/test.dart';

import 'cli.dart' as cli;

void main() {
  test('runCli prints version', () async {
    final stdout = StringBuffer();
    final stderr = StringBuffer();

    final code = await cli.main(<String>['version'], stdoutSink: stdout, stderrSink: stderr);

    expect(code, equals(0));
    expect(stdout.toString().trim(), equals(cli.version));
    expect(stderr.toString(), isEmpty);
  });

  test('runCli prints help', () async {
    final stdout = StringBuffer();
    final stderr = StringBuffer();

    final code = await cli.main(<String>['help'], stdoutSink: stdout, stderrSink: stderr);

    expect(code, equals(0));
    expect(stdout.toString(), contains('usage: gabriel-greeting-dart'));
    expect(stdout.toString(), contains('listLanguages'));
    expect(stderr.toString(), isEmpty);
  });

  test('runCli renders listLanguages as JSON', () async {
    final stdout = StringBuffer();
    final stderr = StringBuffer();

    final code = await cli.main(
      <String>['listLanguages', '--format', 'json'],
      stdoutSink: stdout,
      stderrSink: stderr,
    );

    final payload = jsonDecode(stdout.toString()) as Map<String, dynamic>;
    final languages = payload['languages'] as List<dynamic>;

    expect(code, equals(0));
    expect(languages, hasLength(56));
    expect(languages.first['code'], equals('en'));
    expect(languages.first['name'], equals('English'));
    expect(stderr.toString(), isEmpty);
  });

  test('runCli renders sayHello as text', () async {
    final stdout = StringBuffer();
    final stderr = StringBuffer();

    final code = await cli.main(
      <String>['sayHello', 'Bob', 'fr'],
      stdoutSink: stdout,
      stderrSink: stderr,
    );

    expect(code, equals(0));
    expect(stdout.toString().trim(), equals('Bonjour Bob'));
    expect(stderr.toString(), isEmpty);
  });

  test('runCli defaults sayHello to English JSON output', () async {
    final stdout = StringBuffer();
    final stderr = StringBuffer();

    final code = await cli.main(
      <String>['sayHello', '--json'],
      stdoutSink: stdout,
      stderrSink: stderr,
    );

    final payload = jsonDecode(stdout.toString()) as Map<String, dynamic>;
    expect(code, equals(0));
    expect(payload['greeting'], equals('Hello Mary'));
    expect(payload['language'], equals('English'));
    expect(payload['langCode'], equals('en'));
    expect(stderr.toString(), isEmpty);
  });
}

File: examples/hello-world/gabriel-greeting-node/api/cli_test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');

const cli = require('./cli');

function createBuffer() {
  let contents = '';
  return {
    write(chunk) {
      contents += String(chunk);
    },
    toString() {
      return contents;
    },
  };
}

test('runCLI prints version', async () => {
  const stdout = createBuffer();
  const stderr = createBuffer();

  const code = await cli.runCLI(['version'], stdout, stderr);

  assert.equal(code, 0);
  assert.equal(stdout.toString().trim(), cli.VERSION);
  assert.equal(stderr.toString(), '');
});

test('runCLI prints help', async () => {
  const stdout = createBuffer();
  const stderr = createBuffer();

  const code = await cli.runCLI(['help'], stdout, stderr);

  assert.equal(code, 0);
  assert.match(stdout.toString(), /usage: gabriel-greeting-node/);
  assert.match(stdout.toString(), /listLanguages/);
  assert.equal(stderr.toString(), '');
});

test('runCLI renders listLanguages as JSON', async () => {
  const stdout = createBuffer();
  const stderr = createBuffer();

  const code = await cli.runCLI(['listLanguages', '--format', 'json'], stdout, stderr);
  const payload = JSON.parse(stdout.toString());

  assert.equal(code, 0);
  assert.equal(payload.languages.length, 56);
  assert.equal(payload.languages[0].code, 'en');
  assert.equal(payload.languages[0].name, 'English');
  assert.equal(stderr.toString(), '');
});

test('runCLI renders sayHello as text', async () => {
  const stdout = createBuffer();
  const stderr = createBuffer();

  const code = await cli.runCLI(['sayHello', 'Bob', 'fr'], stdout, stderr);

  assert.equal(code, 0);
  assert.equal(stdout.toString().trim(), 'Bonjour Bob');
  assert.equal(stderr.toString(), '');
});

test('runCLI defaults sayHello to English JSON output', async () => {
  const stdout = createBuffer();
  const stderr = createBuffer();

  const code = await cli.runCLI(['sayHello', '--json'], stdout, stderr);
  const payload = JSON.parse(stdout.toString());

  assert.equal(code, 0);
  assert.equal(payload.greeting, 'Hello Mary');
  assert.equal(payload.language, 'English');
  assert.equal(payload.langCode, 'en');
  assert.equal(stderr.toString(), '');
});
